Sema: Small fixes for result builder inference

AnthonyLatsis opened the pull request:
Fixes #76362, fixes #76746, fixes #76747. 

These are technically source breaking, but the cases in which they break source are quite niche.

AnthonyLatsis committed:
Sema: Result builder inference always favors explicit attr on storage decl

AnthonyLatsis committed:
Sema: Result builder inference from protocol req should look at getter, not storage

AnthonyLatsis committed:
[NFC] Redo and expand result builder inference tests

AnthonyLatsis committed:
Sema: Result builder inference from dynamically replaced decl should look at getter, not storage

AnthonyLatsis commented:
@swift-ci please smoke test Linux

hamishknight commented on lib/Sema/TypeCheckRequestFunctions.cpp:
```diff
@@ -201,30 +201,44 @@ AttachedResultBuilderRequest::evaluate(Evaluator &evaluator,
+  // Check whether there are any return statements in the function's body.
+  // If there are, the result builder transform will be disabled,
+  // so don't infer a result builder.
+  if (!TypeChecker::findReturnStatements(funcDecl).empty())
+    return Type();
```
While here, could change this to use `BraceHasReturnRequest` since that caches the result. Probably also ought to make that request a member on AnyFunctionRef at this point.

AnthonyLatsis replied:
Thanks for the idea!

AnthonyLatsis replied:
> Probably also ought to make that request a member on AnyFunctionRef at this point.

How’s this? I took a slightly different approach.

xedin reviewed (approved):
LGTM! I second Hamish's comment about `BraceHasReturnRequest` it would be good to do at least that here.

AnthonyLatsis committed:
[NFC] AST, Sema: Internalize `BraceHasReturnRequest` evaluation in `AnyFunctionRef` method

AnthonyLatsis committed:
[NFC] AST, Sema: Move `TypeChecker::findReturnStatements` into `AnyFunctionRef`

Also rename it to `getExplicitReturnStmts` for clarity and have it
take a `SmallVector` out parameter instead as a small optimization and
to discourage use of this new method as an alternative to
`AnyFunctionRef::bodyHasExplicitReturnStmt`.

AnthonyLatsis committed:
[NFC] Sema: Reimplement `ReturnStmtFinder` as a function with a callback

A callback enables the caller to control whether to abort the walk. This
is handy for `BraceHasExplicitReturnStmtRequest`.

AnthonyLatsis committed:
[NFC] AST: Rename `BraceHasReturnRequest` to `BraceHasExplicitReturnStmtRequest`

AnthonyLatsis commented:
@swift-ci please smoke test Linux

AnthonyLatsis commented:
@swift-ci please smoke test Linux

AnthonyLatsis commented:
@swift-ci please test source compatibility release

hamishknight commented on lib/Sema/TypeChecker.h:
```diff
@@ -468,8 +468,7 @@ void typeCheckASTNode(ASTNode &node, DeclContext *DC,
 
-/// Find the return statements within the body of the given function.
-std::vector<ReturnStmt *> findReturnStatements(AnyFunctionRef fn);
+bool typeCheckClosureBody(ClosureExpr *closure);
```
Bad rebase? This function doesn't exist anymore

AnthonyLatsis replied:
Good catch.

hamishknight commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -1126,8 +1124,7 @@ ConstraintSystem::matchResultBuilder(AnyFunctionRef fn, Type builderType,
   // https://github.com/apple/swift/issues/64332.
-  if (evaluateOrDefault(getASTContext().evaluator,
-                        BraceHasReturnRequest{fn.getBody()}, false)) {
+  if (fn.getBody()->hasExplicitReturnStmt(getASTContext())) {
```
One advantage of sticking the member on the AnyFunctionRef is you wouldn't have to pass along the ASTContext, but I don't feel too strongly about it

AnthonyLatsis replied:
Yeah, that is unfortunate, and storing the context in the `BraceStmt` just for this sounds silly. 

Discoverability was my main concern with declaring it on `AnyFunctionRef`. Would you mind if I add forwarding methods to `AbstractFunctionDecl` and `AbstractClosureExpr` then?

hamishknight replied:
Sure, sounds good to me!

hamishknight commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -918,13 +918,11 @@ std::optional<BraceStmt *>
   auto &ctx = func->getASTContext();
-  if (evaluateOrDefault(ctx.evaluator, BraceHasReturnRequest{func->getBody()},
-                        false)) {
+  if (SmallVector<ReturnStmt *> returnStmts;
+      func->getBody()->getExplicitReturnStmts(ctx, returnStmts),
+      !returnStmts.empty()) {
```
Cute, but I would rather just bind `returnStmts` outside the `if`, up to you though

AnthonyLatsis committed:
[NFC] Sema: Add more comments to `inferResultBuilderType`

AnthonyLatsis commented:
@swift-ci please test source compatibility release

AnthonyLatsis commented:
@swift-ci please smoke test
